File: 2022/go/02/02.go
package day02

import (
	"fmt"

	"github.com/samber/lo"
)

func scoreLine(line string) int {
	f, s := line[0]-'A', line[2]-'X'
	return int(((s-f+4)%3)*3 + 1 + s)
}

func scoreLine2(line string) int {
	f, s := line[0]-'A', line[2]-'X'
	return int(((f + s + 2) % 3) + 1 + s*3)
}

func Part1(lines []string) {
	score := lo.SumBy(lines, scoreLine)
	fmt.Println(score)
}

func Part2(lines []string) {
	score := lo.SumBy(lines, scoreLine2)
	fmt.Println(score)
}

File: 2023/02/02.fs
﻿module AoC2023.Day02

let lines = System.IO.File.ReadLines "02/data"

let parseCube (cube:string) = match cube.Split(" ") with
                              | [| n; "red" |] -> [|(int n); 0; 0|]
                              | [| n; "green" |] -> [|0; (int n); 0|]
                              | [| n; "blue" |] -> [|0; 0; (int n)|]
                              | _ -> [|0; 0; 0|]
let parseDraw (draw:string) = draw.Split(",", System.StringSplitOptions.TrimEntries) 
                              |> Seq.map parseCube
                              |> Seq.fold (Seq.map2 (+)) [|0;0;0|]
                              |> Seq.toArray

let parseGame (game:string) = (game.Split(":")[1]).Split(";", System.StringSplitOptions.TrimEntries) 
                              |> Seq.map parseDraw
                              |> Seq.toArray

let possible draw = draw |> Seq.fold2 (fun s m d -> s && m >= d) true [|12;13;14|]
let part1 () = lines |> Seq.map parseGame  |> Seq.mapi (fun i game -> if (Seq.forall possible game) then i + 1 else 0) |> Seq.sum |> printfn "%i"

let minPossible =  Seq.fold (Seq.map2 max) [|0;0;0|]
let part2 () = lines |> Seq.sumBy (parseGame >> minPossible >> Seq.fold (*) 1) |> printfn "%i"
